feat: trial table [DET-3019]

hkang1 opened the pull request:
## Description

UPDATED:
<img width="1314" alt="Screen Shot 2020-08-17 at 3 35 54 PM" src="https://user-images.githubusercontent.com/220971/90446911-7379b780-e09f-11ea-94b6-8475287f0af0.png">


## Test Plan



## Commentary (optional)


<!-- User-facing API changes need the "User-facing API Change" label -->

<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:
- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:
- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:
- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")

## Description

Lead with the intended commit body in this description field. For breaking changes, please include "BREAKING CHANGE:" at the beginning of your commit body. At minimum, this section should include a bracketed reference to the Jira ticket, e.g. "[DET-1234]". When squash-and-merging, copy this directly into the description field.

## Test Plan

Describe the situations in which you've tested your change, and/or a screenshot as appropriate. Reviewers may ask questions of this test plan to ensure adequate manual coverage of changes.

## Commentary (optional)

Use this section of your description to add context to the PR. Could be for particularly tricky bits of code that could use extra scrutiny, historical context useful for reviewers, etc. You may intentionally leave this section blank and remove the title.
--->


[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

hkang1 commented on webui/react/src/utils/trial.ts:
```diff
@@ -1,5 +1,13 @@
-import { isNumber } from 'utils/data';
+import { alphanumericSorter, isNumber } from 'utils/data';
+
+export const extractMetricValue = (step: Step, metricName: string): number | undefined => {
```
great call, did not think about the same name being used for both types of metric!

hkang1 commented on webui/react/src/pages/TrialDetails.tsx:
```diff
@@ -1,44 +1,99 @@
+          <Button
+            aria-label="View Checkpoint"
+            icon={<Icon name="checkpoint" />}
+            onClick={e => handleCheckpointShow(e, checkpoint)} />
```
This is a great topic to talk about during WebUI-AG. Thinking something like this for the general component content order?
```
const Component = (props) => {
  ...context api hooks (useStateContext, useActionContext)
  ...useStorage
  ...useStates
  ...useRestApis

  ...useMemos

  ...useEffects

  ...useLayoutEffects

  ...useCallbacks

  if (...) return JSX;
  else if (...) return JSX;
  return (JSX);
};
```

hamidzr replied:
yeah I agree some order is good. although maybe we'd keep component logic small enough by creating more component then it wouldn't be as important but that's more work

hkang1 commented:
> This is looking great and everything works. As we talked there might be an issue with duplicate metric names between training and validation and some of the logic. I added some comments.
> 
> could we either limit the precision of these numbers (accompanying a tooltip) or use the humanizedNumber helper? I'm not sure what AMLE would prefer.
> 

Talked with some AMLE folks and it seems like scientific notation is commonplace. Updated to use the `humanReadableFloat` (scientific notation) plus a tooltip to show the original value per your suggestion.

> * it seems to be missing the has checkpoint filter
> 

Oops forgot this filter, it should be in now.

hkang1 commented on webui/react/src/utils/data.ts:
```diff
@@ -90,6 +90,21 @@ export const commandStateSorter = (a: CommandState, b: CommandState): number =>
+ * Within each type of metric, sort in the order they appear in the `MetricNames` array.
+ * Within the respective type of metrics, `MetricNames` is currently sorted alphanumerically.
+ */
+export const metricNameSorter = (metricNames: MetricNames) => (a: string, b: string): number => {
```
The changes of having a `MetricType` (`Validation` or `Training`) should now be able to distinguish metrics with the same name.

hkang1 commented on webui/react/src/components/SelectFilter.tsx:
```diff
@@ -7,8 +7,11 @@ import Label from './Label';
+  defaultValue?: SelectValue;
   enableSearchFilter?: boolean;
   label: string;
+  mode?: 'multiple' | 'tags';
```
I was hoping to find something like what we had in Elm, but this one is lacking in `antd`. We would have to customize this ourselves.

hkang1 commented:
> looks great!
> question: wdy think about using a smaller label for showing validation vs training type? maybe `t` or `v` with a tooltip? it should be something we'd use a lot in the UI. if we think a smaller label or icon could be as informative to the user with minimal training I think that makes the view cleaner

Love the idea of the smaller tag with a tooltip!

hamidzr commented on webui/react/src/components/SelectFilter.tsx:
```diff
@@ -1,29 +1,28 @@
 import css from './SelectFilter.module.scss';
 
-interface Props {
+interface Props<T = SelectValue> extends SelectProps<T> {
```
:+1:

hamidzr commented:
> Love the idea of the smaller tag with a tooltip!

: ) I'm thinking we might use color os use a bit lighter background for the label later
